src/Guide/Api/Utils.hs

```diff
@@ -89,3 +102,47 @@ instance (HasSwagger api, KnownSymbol name, KnownSymbol desc) =>
              symbolVal (Proxy @desc))
             Nothing
      in toSwagger (Proxy @api) & applyTags [tag]
+
+-- | Servant request details, can be captured by adding `RequestDetails :>` to an API branch.
+data RequestDetails = RequestDetails
+  { rdIp        :: Maybe IP  -- request ip address
+  , rdReferer   :: Maybe Text  -- request referrer
+  , rdUserAgent :: Maybe Text  -- request User-Agent
+  } deriving (Show, Generic)
+
+instance FromHttpApiData IP where
+  parseUrlPiece text
+    | Just ip <- readMaybe $ toString text = Right ip
+    | otherwise = Left "Not a parseable IP address"
+
+instance (HasSwagger api) => HasSwagger (RequestDetails :> api) where
+  toSwagger _ = toSwagger (Proxy :: Proxy api)
+
+instance (HasServer api context)
+  => HasServer (RequestDetails :> api) context where
+
+  type ServerT (RequestDetails :> api) m = RequestDetails -> ServerT api m
+
+  hoistServerWithContext _ pc nt s = hoistServerWithContext (Proxy :: Proxy api) pc nt . s
+
+  route Proxy context subserver = route (Proxy :: Proxy api) context $
+      subserver `addHeaderCheck` withRequest getRequestDetails
+    where
+      getRequestDetails :: Request -> DelayedIO RequestDetails
+      getRequestDetails req = pure $ RequestDetails ip referer useragent
+        where
+          ip :: Maybe IP
+          ip = maybe ipFromHost Just $ (dropEither "Forwarded-For") <|> (dropEither "X-Forwarded-For")
```

Comment:
Try this (not sure if it will work):

```suggestion
          ip = ipFromHost <|> dropEither "Forwarded-For" <|> dropEither "X-Forwarded-For"
```

Reply:
Also, it's not enough to just `read` the IP here. I showed this piece of code before:

```haskell
        addr = T.strip . snd . T.breakOnEnd "," $ forwardedForHeader

        ip = if
           -- [IPv6]:port
           | T.take 1 addr == "[" -> T.drop 1 (T.takeWhile (/= ']') addr)
           -- IPv4 or IPv4:port
           | T.any (== '.') addr  -> T.takeWhile (/= ':') addr
           -- IPv6 without port
           | otherwise            -> addr
```

I.e. you should create an additional helper for parsing the IP out of the `Forwarded-For` header.